core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerProxy.scala

```diff
@@ -820,6 +907,98 @@ object ContainerProxy {
   }
 }
 
+object TCPPingClient {
+  def props(config: ContainerProxyHealthCheckConfig, remote: InetSocketAddress, replies: ActorRef) =
+    Props(new TCPPingClient(remote, replies, config))
+}
+
+class TCPPingClient(remote: InetSocketAddress, listener: ActorRef, config: ContainerProxyHealthCheckConfig)
+    extends Actor {
+  implicit val logging = new AkkaLogging(context.system.log)
+  import context.system
+  implicit val ec = context.system.dispatcher
+  implicit var healthPingTx = TransactionId.actionHealthPing
+
+  var scheduledPing: Option[Cancellable] = None
+  var failedCount = 0
+  var pingMarker: StartMarker = null
+  val tcp = IO(Tcp)
+  var resume: Option[Promise[Unit]] = None
+  val addressString = s"${remote.getHostString}:${remote.getPort}"
+  restartPing()
+
+  private def restartPing() = {
+    cancelPing() //just in case restart is called twice
+    scheduledPing = Some(
+      context.system.scheduler.schedule(config.checkPeriod, config.checkPeriod, self, HealthPingSend(false)))
```

Comment:
Should the initial delay be "0" to send the ping in the Ready state?
https://github.com/apache/openwhisk/pull/4698/files#diff-170ae706abad4ff6500ae12be05d0cbbR518

Reply:
I think waiting at least one checkPeriod before checking the first time is fine, since presumably it was just (immediately prior to this) being used, so we don't need the check till we have stopped using it for a bit. If this is important to be 0, or some other value besides checkPeriod, we could expose another config field to allow operators to specify the initial delay. WDYT?

Reply:
I think I misunderstood the behavior.

It seems it enables Ping when the state become `Ready`.
And after 50ms, it will become `Pausing` and `Paused` state if there is no more activation message. Then it would disable the Ping.
If `checkPeriod` is bigger than 50ms, can it send a ping before the state become `Pausing`?

And when it receives a message and become `Running` again, it would anyway disable the PingActor again. Then can we ping the container before running `/run`?

Reply:
@style95 This test already tests what happens (nothing) when the checkPeriod is bigger than pause timeout https://github.com/apache/openwhisk/pull/4698/files#diff-7731f505e6c9b7417c5b0678a2584fd2R1148 
We only ping between Ready and Pausing - not when Paused

When it becomes Running, the first message will be rescheduled on connection failure, unrelated to health ping.

Reply:
After one invocation is over, it will enable the ping actor by this:
https://github.com/apache/openwhisk/pull/4698/files#diff-170ae706abad4ff6500ae12be05d0cbbR526

If there is no subsequent activation, the container proxy will be `Pausing` state and the ping actor will be disabled again in 50ms(pause grace) by this:
https://github.com/apache/openwhisk/pull/4698/files#diff-170ae706abad4ff6500ae12be05d0cbbR532

Since the initial delay is not 0, no health check will be performed and it will setup `scheduledPing` and disable it immediately(50ms) without any real call.

If the ping actor is only for prewarm containers, do we need to enable the ping actor whenever it becomes a ready state?
Or are you assuming a bigger pause grace than check interval?

Reply:
I understand - I have disabled the checking between Ready and Pausing states, so checking only occurs:
* between Started and Running
* on first activation after Paused (this one is not a ping check, but an HTTP failure will cause activation to be rescheduled instead of failed)